server/src/main/java/org/elasticsearch/indices/breaker/HierarchyCircuitBreakerService.java

```diff
@@ -551,14 +566,18 @@ static class G1OverLimitStrategy implements OverLimitStrategy {
             LongSupplier gcCountSupplier,
             LongSupplier timeSupplier,
             long minimumInterval,
-            TimeValue lockTimeout
+            long fullGCMinimumInterval,
+            TimeValue lockTimeout,
+            TimeValue fullGCLockTimeout
         ) {
             this.lockTimeout = lockTimeout;
+            this.fullGCLockTimeout = fullGCLockTimeout;
             assert minimumInterval > 0;
             this.currentMemoryUsageSupplier = currentMemoryUsageSupplier;
             this.gcCountSupplier = gcCountSupplier;
             this.timeSupplier = timeSupplier;
             this.minimumInterval = minimumInterval;
+            this.fullGCMinimumInterval = fullGCMinimumInterval;
```

Comment:
What's the reason for a separate interval/lock for full gc? Shouldn't it be within the bounds of the existing lock? eg we don't want to try triggering another soft gc right after triggering a full gc?

Reply:
Both are based on suggestions made in the discussion thread: https://elastic.slack.com/archives/CC89MBSTA/p1700197856288609. 

The different interval is pretty straightforward: we want a longer interval between GCs to not have a full GC every 500ms for a long time. 
5s was the original interval for trying to force a young GC. That proved to be too little, so we lowered that to 500ms in https://github.com/elastic/elasticsearch/pull/102396. An interval of 500ms between full GCs seems to be too aggressive, and the original interval was a good starting point.
In general, we want to be able to tune them separately.

The reason to have a separate lock timeout was to leave "more room" for waiting threads if a full GC is happening. This was also one of the suggestions made in the discussion thread. The reason is that a full GC will typically take longer than a young GC; the timeout is there to prevent blocking allocating threads for too long, but in case of full GC we want the timeout to be higher. 
In the end, this is accomplished by re-trying to get the lock for a second time, with the same lock timeout (see previous comments), but again we want to keep that separate in case we want/need to tune them differently.